bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java

```diff
@@ -0,0 +1,154 @@
+                .registerTypeAdapter(OnOffType.class,
+                        (JsonDeserializer<OnOffType>) (json, type,
+                                jsonDeserializationContext) -> json.getAsBoolean() ? OnOffType.ON : OnOffType.OFF)
+                .create();
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing VolvoOnCall API bridge handler.");
+        VolvoOnCallBridgeConfiguration configuration = getConfigAs(VolvoOnCallBridgeConfiguration.class);
+
+        try {
+            httpHeader.setProperty("Authorization", configuration.getAuthorization());
+            try {
+                customerAccount = getURL(SERVICE_URL + "customeraccounts/", CustomerAccounts.class);
+                if (customerAccount.username != null) {
+                    updateStatus(ThingStatus.ONLINE);
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "Incorrect username or password");
+                }
+            } catch (IOException | JsonSyntaxException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            }
+        } catch (UnsupportedEncodingException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+        }
+
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("VolvoOnCall Bridge is read-only and does not handle commands");
+    }
+
+    public String[] getVehiclesRelationsURL() {
+        if (customerAccount != null) {
+            return customerAccount.accountVehicleRelationsURL;
+        } else {
+            return new String[0];
+        }
+    }
+
+    public <T> T getURL(Class<T> objectClass, String vin) throws IOException, JsonSyntaxException {
+        String url = SERVICE_URL + "vehicles/" + vin + "/" + objectClass.getSimpleName().toLowerCase();
+        return getURL(url, objectClass);
+    }
+
+    public <T> T getURL(String url, Class<T> objectClass) throws IOException, JsonSyntaxException {
+        String jsonResponse = HttpUtil.executeUrl("GET", url, httpHeader, null, JSON_CONTENT_TYPE, REQUEST_TIMEOUT);
+        return gson.fromJson(jsonResponse, objectClass);
+    }
+
+    public boolean postURL(String URL, @Nullable String body)
+            throws IOException, InterruptedException, JsonSyntaxException {
+        InputStream inputStream = body != null ? new ByteArrayInputStream(body.getBytes(StandardCharsets.UTF_8)) : null;
+        String jsonString = HttpUtil.executeUrl("POST", URL, httpHeader, inputStream, null, REQUEST_TIMEOUT);
+        PostResponse postResponse = gson.fromJson(jsonString, PostResponse.class);
+        if (postResponse.errorLabel == null) {
+            while (postResponse.status == Status.STARTED || postResponse.status == Status.DELIVERED) {
+                Thread.sleep(1000);
```

Comment:
**Important**: Can you please rewrite the code so the sleep is not needed, the scheduler would be a good start.

This will just a block a thread and bindings should not do this.

Reply:
Can you give me a hint on this one. I can not see how to launch a runnable with a conditional exit ?

Reply:
What do you think of this :  
```        if (postResponse.errorLabel == null) {
            scheduler.scheduleAtFixedRate(() -> {
                try {
                    PostResponse response = getURL(postResponse.serviceURL, PostResponse.class);
                    if (response.status == Status.SUCCESSFULL) {
                        logger.debug("Success !");
                        Thread.currentThread().interrupt();
                    } else {
                        logger.debug("Loop finished with status {}", postResponse.status);
                    }
                } catch (JsonSyntaxException | IOException e) {
                    logger.warn("Error calling API : {}", e.getMessage());
                    Thread.currentThread().interrupt();
                }
            }, 1000, 1000, TimeUnit.MILLISECONDS);
        } else {
            logger.warn(postResponse.errorDescription);
        }```

Reply:
Why do you interrupt? You can save the handler of `scheduleAtFixedRate` and stop the schedule at any time.

Reply:
You could also schedule a single run and reschedule when needed. Also we prefer scheduledWithFixedDelay if that suffices, it puts less burden on the scheduler

Reply:
Yes, but I need to know the result of getURL out of the runnable in order to stop the schedule, this I do not see how to do it differently

Reply:
I don't get your problem tbh.

```java
void myUrlCaller() {
   var response = getURL(...);
   if (response==Status.SUCCESSFULL) { return; }
   // No success, reschedule
   this.handler = scheduler.schedule(this::myUrlCaller, 1000, TimeUnit.Milliseconds);
}

void somewhere() {
   if (this.handler) this.handler.cancel();
   this.handler = scheduler.schedule(this::myUrlCaller, 1000, TimeUnit.Milliseconds);
}
```

Reply:
Well, in fact this is an heavier refactoring because it implies desynchronizing action from feed-back with bigger impacts. Could this not be accepted this way in a first step ?

Reply:
thread sleep calls cannot be accepted. There is only exception, if you need timing critical rs232 output or similar and the protocol requires a 20ms pulse etc.

Just wrap your head around the concept of the scheduler. You haven't understood it yet, judging by your responses. "desynchronizing action from feed-back" whatever you mean in here, I bet it's not true :)